### DISCUSSION TITLE
How to access per device advertisment data?

### DISCUSSION THREAD

**Discussion by zapta** [hbldh/bleak#1271] [Q&A]
I have simple code that scans for 5 secs and then prints the devices it found. I want to extend it to print some per device advertisement data. I tried device.metadata but it seems to be deprecated, with some reference to a callback which I don't use. What is the new way of doing so?

https://github.com/zapta/ble_stepper_motor_analyzer/blob/main/python/common/connections.py#L29

**Comment by dlech** [comment]
https://bleak.readthedocs.io/en/latest/api/scanner.html#bleak.BleakScanner.discovered_devices_and_advertisement_data

**Comment by zapta** [comment]
Thanks @dlech. I am not sure how to use it. I tried to add the print statement at the bottom, expecting to get a dump of a dictionary keyed by device id, but got instead 'property'.  My goal is to print in the loop, for each device, a value of a specific advertisement attribute (manufacturer data). Any suggestion on how to do that? Is it even possible with the 'easy method' ?

```python
async def scan_and_dump():
    print("Scanning 5 secs for advertising BLE devices ...", flush=True)
    devices = await BleakScanner.discover(timeout=5)
    i = 0
    for device in devices:
        # print(f"Metadata: {device.}", flush=True)
        i += 1
        # print(device, flush=True)
        name = device.name or ""
        print(f"{i:2} device address: {device.address}  ({name})", flush=True)
        
    print(f"Advertisement data:\n{BleakScanner.discovered_devices_and_advertisement_data}\n", flush=True)
```

Output, 
```
 1 device address: 7B89D6D3-4612-8D76-6278-9A31CC6C9DAD  (STP-0C8B95F3089E)
 2 device address: F1712D95-6D4D-05F9-8A26-0BE5D90DE2BF  ()
 3 device address: 9BAEB8A6-E88B-4B4B-1F1B-529C5141D620  (STP-0C8B95F2B436)
 ...

Advertisement data:
<property object at 0x124d42cf0>

```

**Reply by dlech** [reply]
Add `return_adv=True` to `discover()` and change `for device in devices:` to `for device, adv in devices.values():`

https://bleak.readthedocs.io/en/latest/api/scanner.html#bleak.BleakScanner.discover

**Comment by zapta** [comment]
Thanks @dlech, I do get it now. 

Is the rssi value the actual signal strength or some fixed adv data?  If it is, is it available on all three platforms, Win, Linux, Mac?


```
AdvertisementData(local_name='STP-0C8B95F2B436', manufacturer_data={4369: b'XYZ'}, tx_power=9, rssi=-63)
```

**Reply by dlech** [reply]
It is the last value received. It is available on all platforms.

**Comment by zapta** [comment]
Very good. Thanks.
